dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/RuleActionScheduleMessageImplementer.java

```diff
@@ -52,144 +47,80 @@
-      handleSingleEvent(ruleEffect, event);
-      return;
-    }
-
-    if (!result.isValid()) {
-      return;
-    }
-
-    ProgramNotificationTemplate template = result.getTemplate();
-    String key = generateKey(template, event.getEnrollment());
+    ProgramNotificationTemplate template =
+        notificationHelper.getNotificationTemplate(ruleEffect.getRuleAction());
 
     String date = StringUtils.unwrap(ruleEffect.getData(), '\'');
 
-    if (!isValid(date)) {
-      return;
-    }
-
-    ProgramNotificationInstance notificationInstance =
-        notificationTemplateService.createNotificationInstance(template, date);
-    notificationInstance.setEvent(event);
-    notificationInstance.setEnrollment(null);
-
-    programNotificationInstanceService.save(notificationInstance);
-
-    log.info(String.format(LOG_MESSAGE, template.getUid()));
-
-    if (result.getLogEntry() != null) {
+    if (isInvalid(date)) {
       return;
     }
 
-    ExternalNotificationLogEntry entry = createLogEntry(key, template.getUid());
-    entry.setNotificationTriggeredBy(NotificationTriggerEvent.PROGRAM_STAGE);
-    entry.setAllowMultiple(template.isSendRepeatable());
+    NotificationValidationResult result =
+        notificationHelper.validate(template, event.getEnrollment());
 
-    notificationLoggingService.save(entry);
-  }
+    if (result.isValid()) {
+      ProgramNotificationInstance notificationInstance =
+          notificationTemplateService.createNotificationInstance(template, date);
+      notificationInstance.setEvent(event);
+      notificationInstance.setEnrollment(null);
 
-  // -------------------------------------------------------------------------
-  // Supportive Methods
-  // -------------------------------------------------------------------------
+      programNotificationInstanceService.save(notificationInstance);
 
-  private void handleSingleEvent(RuleEffect ruleEffect, Event event) {
-    ProgramNotificationTemplate template = getNotificationTemplate(ruleEffect.getRuleAction());
+      log.info(String.format(LOG_MESSAGE, template.getUid()));
 
-    if (template == null) {
-      return;
+      if (result.needsToCreateLogEntry()) {
+        notificationHelper.createLogEntry(template, event.getEnrollment());
+      }
     }
-
-    String date = StringUtils.unwrap(ruleEffect.getData(), '\'');
-
-    if (!isValid(date)) {
-      return;
-    }
-
-    ProgramNotificationInstance notificationInstance =
-        notificationTemplateService.createNotificationInstance(template, date);
-    notificationInstance.setEvent(event);
-    notificationInstance.setEnrollment(null);
-
-    programNotificationInstanceService.save(notificationInstance);
-
-    log.info(String.format(LOG_MESSAGE, template.getUid()));
   }
 
-  private boolean isValid(String date) {
-    return StringUtils.isNotBlank(date) && DateUtils.dateIsValid(date);
+  private boolean isInvalid(String date) {
+    return StringUtils.isBlank(date) || !DateUtils.dateIsValid(date);
```

Comment:
user can specify the date as `'2024-12-12'`.

Reply:
I am not sure what you want to say by that.
Anyway, the behavior is exactly the same as before